---
title: "RNA SBS Plotting"
author: "BF"
date: "2024-07-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
# Install SomaticSignatures package
#BiocManager::install("SomaticSignatures")

#install.packages("ggforce")
```


```{r}
# Load necessary libraries
library(SomaticSignatures)
library(BSgenome)
library(ggplot2)
library(ggforce)
library(tidyverse)
```


```{r}
# Define the path to signatures file
file_path <- "/Users/benfixman/Library/Mobile Documents/com~apple~CloudDocs/Chen_Lab/RNA_SBS/Transient overexpression of exogenous APOBEC3A causes C-to-U RNA editing of thousands of genes/SigProfiler Extractor/Published Variants/Sigprofiler Extractor/CH192/Suggested_Solution/CH192_De-Novo_Solution/Signatures/CH192_De-Novo_Signatures.txt"

# Read the data into R
signatures <- read.table(file_path, header = TRUE, sep = "\t", row.names = 1)

# Display the first few rows of the data to ensure it's loaded correctly
head(signatures)

# Convert the data to long format
signatures_long <- signatures %>%
  rownames_to_column("Mutation") %>%
  pivot_longer(-Mutation, names_to = "Signature", values_to = "Contribution")

# Add a column for mutation categories and context
signatures_long <- signatures_long %>%
  mutate(Mutation_Category = substr(Mutation, 3, 5),
         Simple_Mutation = paste0(substr(Mutation, 1, 1), substr(Mutation, 3, 3), substr(Mutation, 7, 7)))

# Define custom colors for the mutation categories
custom_colors <- c("C>A" = "#00AEEF", "C>G" = "#000000", "C>T" = "#FF0000", 
  "T>A" = "#A9A9A9", "T>C" = "#008000", "T>G" = "#FF69B4",
  "G>T" = "#ADD8E6", "G>C" = "#808080", "G>A" = "#FFA500",
  "A>T" = "#D3D3D3", "A>G" = "#E6F5B0", "A>C" = "#FFB6C1")

# Define the order of mutation categories
mutation_order <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G", "G>T", "G>C", "G>A", "A>T", "A>G", "A>C")

# Convert Mutation_Category to a factor with the specified order
signatures_long <- signatures_long %>%
  mutate(Mutation_Category = factor(Mutation_Category, levels = mutation_order))

# plot code
p <- ggplot(signatures_long, aes(x = Simple_Mutation, y = Contribution, fill = Mutation_Category)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Mutation_Category, scales = "free_x", nrow = 1) +
  scale_fill_manual(values = custom_colors) +
  scale_y_continuous(breaks = seq(0, 0.4, by = 0.1), expand = c(0, 0)) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 0.5, size = 6), # Rotate and adjust text size
    axis.title.x = element_blank(), # Remove the x-axis title
    axis.title.y = element_text(size = 12), # Adjust y-axis title size
    legend.position = "none", # Remove legend
    strip.text = element_text(size = 10), # Adjust facet label size
    strip.background = element_rect(fill = NA, color = NA), # Remove default facet background
    panel.grid = element_blank(), # Remove grid lines
    panel.background = element_rect(fill = "white", color = NA), # Set background to white
    axis.line = element_line(size = 0.5, color = "black"), # Add x and y axis lines
    axis.ticks = element_line(size = 0.2, color = "black"), # Add tick marks
    axis.ticks.length = unit(0.1, "cm") # Adjust tick mark length
  ) +
  labs(y = "% SBS")

# Save the plot with adjusted dimensions and white background
ggsave("A3A_Exogenous_mutational_signature_plot.png", plot = p, width = 16, height = 4, units = "in", bg = "white")
```


```{r}
# Define the path to signatures file
file_path <- "/Users/benfixman/Library/Mobile Documents/com~apple~CloudDocs/Chen_Lab/RNA_SBS/APOBEC3A cytidine deaminase induces RNA editing in monocytes and macrophages/Data/a3a-polarization-extractor/CH192/Suggested_Solution/CH192_De-Novo_Solution/Signatures/CH192_De-Novo_Signatures.txt"

# Read the data into R
signatures <- read.table(file_path, header = TRUE, sep = "\t", row.names = 1)

# Display the first few rows of the data to ensure it's loaded correctly
head(signatures)

# Convert the data to long format
signatures_long <- signatures %>%
  rownames_to_column("Mutation") %>%
  pivot_longer(-Mutation, names_to = "Signature", values_to = "Contribution")

# Add a column for mutation categories and context
signatures_long <- signatures_long %>%
  mutate(Mutation_Category = substr(Mutation, 3, 5),
         Simple_Mutation = paste0(substr(Mutation, 1, 1), substr(Mutation, 3, 3), substr(Mutation, 7, 7)))

# Define custom colors for the mutation categories
custom_colors <- c("C>A" = "#00AEEF", "C>G" = "#000000", "C>T" = "#FF0000", 
  "T>A" = "#A9A9A9", "T>C" = "#008000", "T>G" = "#FF69B4",
  "G>T" = "#ADD8E6", "G>C" = "#808080", "G>A" = "#FFA500",
  "A>T" = "#D3D3D3", "A>G" = "#E6F5B0", "A>C" = "#FFB6C1")

# Define the order of mutation categories
mutation_order <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G", "G>T", "G>C", "G>A", "A>T", "A>G", "A>C")

# Convert Mutation_Category to a factor with the specified order
signatures_long <- signatures_long %>%
  mutate(Mutation_Category = factor(Mutation_Category, levels = mutation_order))

# plot code
p <- ggplot(signatures_long, aes(x = Simple_Mutation, y = Contribution, fill = Mutation_Category)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~Mutation_Category, scales = "free_x", nrow = 1) +
  scale_fill_manual(values = custom_colors) +
  scale_y_continuous(breaks = seq(0, 0.4, by = 0.1), expand = c(0, 0)) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 0.5, size = 6), # Rotate and adjust text size
    axis.title.x = element_blank(), # Remove the x-axis title
    axis.title.y = element_text(size = 12), # Adjust y-axis title size
    legend.position = "none", # Remove legend
    strip.text = element_text(size = 10), # Adjust facet label size
    strip.background = element_rect(fill = NA, color = NA), # Remove default facet background
    panel.grid = element_blank(), # Remove grid lines
    panel.background = element_rect(fill = "white", color = NA), # Set background to white
    axis.line = element_line(size = 0.5, color = "black"), # Add x and y axis lines
    axis.ticks = element_line(size = 0.2, color = "black"), # Add tick marks
    axis.ticks.length = unit(0.1, "cm") # Adjust tick mark length
  ) +
  labs(y = "% SBS")

# Save the plot with adjusted dimensions and white background
ggsave("A3A_endogenous_mutational_signature_plot.png", plot = p, width = 16, height = 4, units = "in", bg = "white")
```












